Improve AccessTokenUnavailableException error message

Recently I was facing an issue where I was having a hard time to understand why an exception being thrown. I had the following scenario:
* _accessTokens_ internal map correctly fetched my token with, say, tokenId "my-app"
* And, due to a type, I was using tokenId parameter as "my-app'" (note the additional single quote)

As I mentioned, I was receiving an exception like:

`
Caused by: org.zalando.stups.tokens.AccessTokenUnavailableException: no token available
	at org.zalando.stups.tokens.AbstractAccessTokenRefresher.getAccessToken(AbstractAccessTokenRefresher.java:83) ~[tokens-0.11.0-beta-2.jar:na]
`

Having a error message like "no token avaiable" is not helpfull at all. 
At that moment in the code, we could add to the error message something like:

_"No token available for tokenId [tokenId]. Tokens are available for the following tokenIds [<list_of_tokenIds_available>]"_

What do you think?
I myself could work on this issue if you all agree.

--- a/src/main/java/org/zalando/stups/tokens/AbstractAccessTokenRefresher.java
+++ b/src/main/java/org/zalando/stups/tokens/AbstractAccessTokenRefresher.java
@@ -15,17 +15,24 @@
  */
 package org.zalando.stups.tokens;
 
+import static java.lang.String.format;
+
 import java.util.Date;
+import java.util.List;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.TimeUnit;
 
+import org.assertj.core.util.Lists;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 public abstract class AbstractAccessTokenRefresher implements AccessTokens {
 
     private static final Logger LOG = LoggerFactory.getLogger(AbstractAccessTokenRefresher.class);
+
+    private static final String NOT_AVAILABLE_TEMPLATE = "No token available for tokenId '%s'. Tokens are available for the following tokenIds %s";
+    private static final String EXPIRED_TEMPLATE = "AccessToken for tokenId '%s' expired.";
 
     private static final long ONE_YEAR_SECONDS = TimeUnit.DAYS.toSeconds(365);
     private static final String FIXED_TOKENS_ENV_VAR = "OAUTH2_ACCESS_TOKENS";
@@ -34,6 +41,8 @@ public abstract class AbstractAccessTokenRefresher implements AccessTokens {
     protected final ScheduledExecutorService scheduler;
 
     protected final ConcurrentHashMap<Object, AccessToken> accessTokens = new ConcurrentHashMap<>();
+
+    private String availableTokenIds;
 
     public AbstractAccessTokenRefresher(TokenRefresherConfiguration configuration) {
         this.configuration = configuration;
@@ -80,14 +89,25 @@ public abstract class AbstractAccessTokenRefresher implements AccessTokens {
     public AccessToken getAccessToken(final Object tokenId) throws AccessTokenUnavailableException {
         final AccessToken token = accessTokens.get(tokenId);
         if (token == null) {
-            throw new AccessTokenUnavailableException("no token available");
+            throw new AccessTokenUnavailableException(format(NOT_AVAILABLE_TEMPLATE, tokenId.toString(), getAvailableTokenIds()));
         }
 
         if (token.isExpired()) {
-            throw new AccessTokenUnavailableException("token expired");
+            throw new AccessTokenUnavailableException(format(EXPIRED_TEMPLATE, tokenId.toString()));
         }
 
         return token;
+    }
+
+    protected String getAvailableTokenIds() {
+        if(availableTokenIds == null) {
+            List<String> tokenIds = Lists.newArrayList();
+            accessTokens.forEachKey(Long.MAX_VALUE, key -> {
+                tokenIds.add(key.toString());
+            });
+            availableTokenIds = tokenIds.isEmpty() ? "[]" : tokenIds.toString();
+        }
+        return availableTokenIds;
     }
 
     @Override

--- a/src/main/java/org/zalando/stups/tokens/AbstractAccessTokenRefresher.java
+++ b/src/main/java/org/zalando/stups/tokens/AbstractAccessTokenRefresher.java
@@ -17,13 +17,13 @@ package org.zalando.stups.tokens;
 
 import static java.lang.String.format;
 
+import java.util.ArrayList;
 import java.util.Date;
 import java.util.List;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.TimeUnit;
 
-import org.assertj.core.util.Lists;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -101,7 +101,7 @@ public abstract class AbstractAccessTokenRefresher implements AccessTokens {
 
     protected String getAvailableTokenIds() {
         if(availableTokenIds == null) {
-            List<String> tokenIds = Lists.newArrayList();
+            List<String> tokenIds = new ArrayList<String>();
             accessTokens.forEachKey(Long.MAX_VALUE, key -> {
                 tokenIds.add(key.toString());
             });

--- a/src/test/java/org/zalando/stups/tokens/AbstractAccessTokenRefresherTest.java
+++ b/src/test/java/org/zalando/stups/tokens/AbstractAccessTokenRefresherTest.java
@@ -1,0 +1,60 @@
+package org.zalando.stups.tokens;
+
+import java.util.Arrays;
+import java.util.List;
+
+import org.assertj.core.api.Assertions;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.contrib.java.lang.system.EnvironmentVariables;
+import org.mockito.Mockito;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class AbstractAccessTokenRefresherTest {
+
+    private static final String FIXED_TOKENS = "one=DSAFDASASDFDA,two=DSFADFADFADFADFA,three=ADAFDAFADFAFEWRDFADFASDF";
+    private static final String EXPECTED_MESSAGE = "No token available for tokenId 'four'. Tokens are available for the following tokenIds [one, two, three]";
+
+    @Rule
+    public final EnvironmentVariables environmentVariables = new EnvironmentVariables();
+
+    @Test
+    public void testListToString() {
+        List<String> keys = Arrays.asList("One", "Two", "Three");
+        Assertions.assertThat(keys.toString()).isEqualTo("[One, Two, Three]");
+    }
+
+    @Test
+    public void testExceptionMessage() {
+        environmentVariables.set("OAUTH2_ACCESS_TOKENS", FIXED_TOKENS);
+        TokenRefresherConfiguration configuration = Mockito.mock(TokenRefresherConfiguration.class);
+        AbstractAccessTokenRefresher refresher = new TestTokenRefresher(configuration);
+        refresher.initializeFixedTokensFromEnvironment();
+        String availableTokenIds = refresher.getAvailableTokenIds();
+        Assertions.assertThat(availableTokenIds).isEqualTo("[one, two, three]");
+        try {
+            refresher.get("four");
+            Assertions.fail("Expect an exception to be thrown because the token 'four' is not available");
+        } catch (AccessTokenUnavailableException e) {
+            String message = e.getMessage();
+            Assertions.assertThat(message).isEqualTo(EXPECTED_MESSAGE);
+        }
+
+    }
+
+    static class TestTokenRefresher extends AbstractAccessTokenRefresher {
+
+        private static final Logger LOG = LoggerFactory.getLogger(TestTokenRefresher.class);
+
+        public TestTokenRefresher(TokenRefresherConfiguration configuration) {
+            super(configuration);
+        }
+
+        @Override
+        public void start() {
+            LOG.info("START {} ....", getClass().getName());
+        }
+
+    }
+}

--- a/src/test/java/org/zalando/stups/tokens/AbstractAccessTokenRefresherTest.java
+++ b/src/test/java/org/zalando/stups/tokens/AbstractAccessTokenRefresherTest.java
@@ -1,3 +1,18 @@
+/**
+ * Copyright (C) 2015 Zalando SE (http://tech.zalando.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
 package org.zalando.stups.tokens;
 
 import java.util.Arrays;

